### PR TITLE
[homematic] Dynamically assign reconfigurable channels

### PR DESCRIPTION
**Pull request by maniac103**
HM-MOD-EM-8 devices have reconfigurable channels, whose data point set (and thus set of available thing channels) varies depending on what type the channel is configured to. For such devices, put the list of available channel types not in the ThingType, but assign it to each Thing when initializing it.

### CODE REVIEW

**Comment by maniac103** [comment]
@gerrieg As discussed in #2930 ... I still need to try whether this actually works as expected, but I figured I'll open the PR already for you to review whether the chosen approach is OK.

One could probably also update the thing channel list dynamically by running the channel list update code via HomematicBridgeHandler::reloadDeviceValues (need to fetch channel values in that case though), but I'm not sure what thread this is run from and/or should be run from.

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/HomematicTypeGeneratorImpl.java [hunk 1]
```diff
@@ -214,7 +214,7 @@ private ThingType createThingType(HmDevice device, List<ChannelGroupType> groupT
         properties.put(Thing.PROPERTY_MODEL_ID, device.getType());
 
         URI configDescriptionURI = getConfigDescriptionURI(device);
-        if (configDescriptionProvider.getConfigDescription(configDescriptionURI, null) == null) {
+        if (configDescriptionProvider.getConfigDescription(configDescriptionURI, Locale.getDefault()) == null) {
```
Comment (maniac103) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/HomematicTypeGeneratorImpl.java]: Not sure whether this is the correct thing to do. `ConfigDescriptionProvider.getConfigDescription()` has a `NonNullByDefault` annotation (which also applies to the locale parameter), but all users of the locale parameter either had no annotation at all or even had a `Nullable` annotation (e.g. `ProfileTypeRegistry.getProfileTypes`)

@kaikreuzer Do you know what the correct direction of fixing this is?

Comment (kaikreuzer) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/HomematicTypeGeneratorImpl.java]: It is correct that the `ConfigDescriptionProvider` interface promises that the framework will always pass a locale when calling this method, so this annotation is good and helpful.

We have here now the (rare) situation that the caller is not the framework, but the binding itself. But as it isn't calling any `ConfigDescriptionProvider`, but a `HomematicConfigDescriptionProvider`, my suggestion is to declare that we know for this one that it is safe to also call it with a null value by adding:

```
public interface HomematicConfigDescriptionProvider extends ConfigDescriptionProvider {

    @Override
    @Nullable
    ConfigDescription getConfigDescription(URI uri, @Nullable Locale locale);
```

This should solve the issue, right?

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/HomematicTypeGeneratorImpl.java [hunk 2]
```diff
@@ -298,7 +303,7 @@ private void generateConfigDescription(HmDevice device, URI configDescriptionURI
         for (HmChannel channel : device.getChannels()) {
             String groupName = "HMG_" + channel.getNumber();
             String groupLabel = MetadataUtils.getDescription("CHANNEL_NAME") + " " + channel.getNumber();
-            groups.add(new ConfigDescriptionParameterGroup(groupName, null, false, groupLabel, null));
+            groups.add(new ConfigDescriptionParameterGroup(groupName, null, false, groupLabel, ""));
```
Comment (kaikreuzer) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/type/HomematicTypeGeneratorImpl.java]: This change actually looks doggy - upon closer analysis, I created https://github.com/eclipse/smarthome/pull/4952, so that you should be able to revert this change again.

**Comment by gerrieg** [comment]
First of all thank you @maniac103 for taking care of the problem. My time is currently very limited and i also can not test it because i don't have this device.
The approach LGTM. But don't call HomematicBridgeHandler::reloadDeviceValues, you have to use triggerDeviceValuesReload from AbstractHomematicGateway to correctly reload the values.

**Comment by maniac103** [comment]
> But don't call HomematicBridgeHandler::reloadDeviceValues, you have to use triggerDeviceValuesReload from AbstractHomematicGateway to correctly reload the values.

The idea is not to call reloadDeviceValues from somewhere, but to call the newly added channel update code from it (so it's done whenever reloadDeviceValues is called).

**Comment by gerrieg** [comment]
Then it should work already. In reloadDeviceValues there is already a call to thingUpdated in the thingHandler. And there the initialize method is called.

**Comment by maniac103** [comment]
Ah, OK, then we should be fine indeed.

**Comment by maniac103** [comment]
Implemented the review comments by Kai. I also tested it now and fixed a few issues I noticed during testing. The dynamic population of channels works now, but dynamic repopulation when changing the HM channel type doesn't work because afterwards the data point set would need to be repopulated, which - AFAICT - doesn't happen right now (only initial population of a new device set happens upon scanning).

@gerrieg Do you have any suggestions about this - where/how should this repopulation of the HM datapoint set for a given device's HM channels take place?

**Comment by gerrieg** [comment]
I can't tell you that by heart, i'd have to test it out first, too.

**Comment by maniac103** [comment]
OK. I'll see what I find and let you know once I have something up for review.

**Comment by maniac103** [comment]
This should do it. I just need to add Javadoc comments.

**Comment by maniac103** [comment]
This is ready from my side now.

**Comment by maniac103** [comment]
This is ready from my side now.

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmChannel.java [hunk 3]
```diff
@@ -142,6 +136,47 @@ public boolean hasDatapoint(HmDatapointInfo dpInfo) {
...
+        return getDatapoint(HmParamsetType.MASTER, HomematicConstants.DATAPOINT_NAME_CHANNEL_FUNCTION) != null;
+    }
+
+    public Integer getCurrentFunction() {
+        HmDatapoint functionDp = getDatapoint(HmParamsetType.MASTER,
+                HomematicConstants.DATAPOINT_NAME_CHANNEL_FUNCTION);
+        return functionDp == null ? null : (Integer) functionDp.getValue();
+    }
+
+    public boolean hasChannelFunctionChanged() {
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmChannel.java]: Imho this name does not suggests a side effect of storing the `lastFunction` I think this potentially a place for bugs to come from.

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDevice.java [hunk 4]
```diff
@@ -38,34 +38,30 @@
 
     private List<HmChannel> channels = new ArrayList<HmChannel>();
 
+    public HmDevice(String address, HmInterface hmInterface, String type, String gatewayId, String homegearId,
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDevice.java]: A constructor with many strings such as this is error prone (too easy to mess up the order of values), I'm in doubt whether this is really a good alternative, wdyt?

Comment (maniac103) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDevice.java]: You mean a good alternative over the setters? If yes, I disagree: the reason why I replaced the setters with the constructor is that it wasn't clear to me at the beginning what class members are 'static' (only set initially and never touched again afterwards) and what members are expected to change during the object's lifetime. By removing the setters and replacing them by constructor parameters, this immediately becomes clear (albeit I probably should make the member variables final now to make it even more clear - will do that once we reach consensus).

One way to improve (at least a bit) would be reordering the parameter list, e.g. putting `hmInterface` between `type` and `gatewayId`.

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmChannel.java (hunk 3 above)
Comment (maniac103) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmChannel.java]: Any alternative suggestions? ;-)

I initially had the comparison in the calling code and had an additional method (`acknowledgeCurrentFunction()`) that updated `lastFunction`, but that solution seemed more fragile to me because it requires the caller to do The Right Thing™. Having only one function to me seemed like the better, more simple solution.

Maybe renaming the method would be an option, something like `checkForChannelFunctionChange()`?

BTW, I should also add Javadoc comments here to document the newly added methods.

Comment (martinvw) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmChannel.java]: Agree on both (name & adding java docs for it)

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDevice.java (hunk 4 above)
Comment (martinvw) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDevice.java]: An alternative which makes your change way bigger is adding real types / enums for `HmDeviceType type, GatewayId gatewayId, HomeGearId homegearId`

If you make sure you validate by hand the refactored code and do some proper testing I'm okay if @gerrieg is okay with this.

Comment (maniac103) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDevice.java]: > An alternative which makes your change way bigger is adding real types / enums for HmDeviceType type, GatewayId gatewayId, HomeGearId homegearId

Unfortunately that doesn't really work. The gateway ID and Homegear ID are just that - (opaque) IDs: gateway ID is the thing ID of the gateway this device is connected to, and Homegear ID is the ID used by Homegear to reference this device. Both can't be meaningfully interpreted. For `type` one could create an enum, however, doing so has disadvantages: type is often used as part of unique ID generation (e.g. for things, thing channels etc.). One could still do this with an enum, but there's a problem when devices emerge that were unknown at binding creation time: one would need to summarize them in an 'unknown' enum value, so the identification function would be lost for those devices, which in all other aspects are likely to work just fine.

If this change is controversial, I can also take it out of this PR as it's not directly related - it was something I just noticed when I made myself familiar with the code base, and other classes (e.g. `HmDatapoint` or `HmDatapointInfo`) already followed the same style.

Comment (martinvw) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmDevice.java]: No as long as @gerrieg still approved he already he did, so without any further messages I assume he still does :-) Its okay I just noted that the new choice has its own drawbacks imho. For me it's not a reason to stop this from being merged.

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmChannel.java (hunk 3 above)
Comment (martinvw) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmChannel.java]: I will wait with merging until you did this, agree? Can you ping me when you finished that?

Comment (maniac103) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/internal/model/HmChannel.java]: Yeah, sure ... @martinvw ping :-)

**Comment by martinvw** [comment]
@gerrieg do you want to do another test / review? Or are you fine with I believe there is a new commit "[homematic] Avoid potential threading issues during thing initialization"

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/HomematicThingHandler.java [hunk 5]
```diff
@@ -111,6 +100,120 @@ public void run() {
         });
     }
 
+    private void doInitializeInBackground()
```
Comment (martinvw) [inline addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/HomematicThingHandler.java]: Small semantic comment this method is not the reason it's running in the background, but it's wrapped a few lines up that is why its running in the background.

Comment (maniac103) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/HomematicThingHandler.java]: That's right. With the method naming I intended to make clear in what scenario the method is supposed to be run in.

**Comment by maniac103** [comment]
> Or are you fine with it, I believe there is a new commit "[homematic] Avoid potential threading issues during thing initialization"

Indeed, that one is new, as I noticed two initialize() methods running simultaneously (scheduled on 2 different scheduler threads), which seemed dangerous to me. As the commit message states this wasn't as much of a problem before, as the datapoint list was initialized in the thread the HmChannel object was initialized in and never changed afterwards. This has changed now though, as datapoints can be changed if channel function changes.

If  @gerrieg has time for a last review round, that'd be good :-)

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/HomematicThingHandler.java (hunk 5 above)
Comment (maniac103) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/HomematicThingHandler.java]: (Having said that, I'm fine with removing the InBackground suffix if you prefer)

File: addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/HomematicThingHandler.java [hunk 6]
```diff
@@ -105,9 +105,9 @@ public void run() {
                     // update configurations
                     Configuration config = editConfiguration();
                     for (HmChannel channel : device.getChannels()) {
+                        loadHomematicChannelValues(channel);
```
[the remark this answers is not in the document]
Comment (maniac103) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/HomematicThingHandler.java]: The thing is, all datapoints (including VALUES) were fetched before as well (assuming all existing devices have at least one MASTER datapoint), but they were fetched _once per MASTER datapoint_. Taking the (random) example of HM-MOD-EM-8: a channel of it has 4 MASTER datapoints, so `loadHomematicChannelValues()` was run 4 times per channel before and is run one time per channel now. The only scenario where there would be more RPC calls than before is devices with channels that don't have any MASTER datapoints - are there any such devices? At least the 4 device types I own (HM-LC-Bl1-FM, HM-MOD-EM-8, HM-MOD-Re-8, HM-LC-Sw1-FM) have MASTER datapoints on all channels.

As you mentioned one could optimize further by only loading MASTER values instead of all values, but that's a larger/more involved change that I didn't feel like making at this point. I wanted to pick the low-hanging fruit first :-)

[the remark this answers is not in the document]
Comment (gerrieg) [reply addons/binding/org.openhab.binding.homematic/src/main/java/org/openhab/binding/homematic/handler/HomematicThingHandler.java]: You are right, this is something i need to optimize in the future. However, only the method was called four times. Due to the ```hmChannel.isInitialized()``` method, RPC call is done only once.

OK, then grab the low hanging fruits now ;-)
